documentation/nearest-neighbor-search.html

```diff
@@ -0,0 +1,436 @@
+using a more advanced <a href="onnx.html">ONNX</a> imported Machine Learned (ML) model. These are examples of multi-stage or <a href="phased-ranking.html">phased ranking</a>.
+</p>
+
+<pre>
+rank-profile image-similarity-with-onnx-stage {
+  first-phase {
+    expression: closeness("field", image_sift_encoding)*attribute(popularity) 
+  }
+  second-phase {
+    rerank-count: 100
+    expression: onnx("image-similarity-model.onnx")
+  }
+ }
+</pre>
+<p>With the above multi-stage ranking profiles there are three stages: 
+<ul>
+<li>The zero-phase nearest neighbor search which only ranks documents by the closeness feature using <em>angular-distance</em></li>
+<li>The <em>first-phase</em> which re-ranks the top <em>targetHits</em> from zero-phase</li>
+<li>the <em>second-phase</em> which re-ranks the top 100 hits from <em>first-phase</em> ranking stage</li>
+</ul>
+<p>
+<b>Related Products using BERT Text Embedding Similarity</b>
+<p>As an alternative to using product image similarity to find related items we look at using the BERT embedding from the product title instead. 
+Similar to the image similarity case we assume that we
+have the BERT embedding tensor representation for the product we want to find related or similar items for 
+and we can use this to search for similar products using the angular distance.  We define our
+ranking function:</p>
+
+<pre>
+rank-profile title-similarity-using-bert {
+  first-phase {
+    expression: closeness("field", title_bert_embedding)
+  }
+}
+</pre> 
+<p>
+Search for 10 nearest neighbors using angular distance in the BERT embedding vector space and we are only interested in products where <em>in_stock</em> is true:
+</p>
+<pre>
+{
+  "yql": "select title,price from sources products where ([{\"targetNumHits\": 10}]nearestNeighbor(title_bert_embedding,query_vector_bert)) and in_stock = true;",
+  "ranking.features.query(query_vector_bert)": [
+    0.7528874147920314,
+    0.712266471788029,
+    0.8813024904379954,
+    ...
+  ],
+  "ranking.profile": "title-similarity-using-bert",
+  "hits": 10
+}
+</pre>
+
+<b>Related Products using hybrid BERT Text Embedding Similarity</b>
+<p> 
+We can also combine and retrieve products in a single query request using disjunction (OR) 
+logical operator to retrieve products which are close in either vector space and we define 
+our ranking function as a weighted linear combination of the closeness scores produced by two nearest neightbor search operators:</p>
+
+<pre>
+rank-profile hybrid-similarity {
+  first-phase {
+    expression: query(bert_weight)*closeness("field", title_bert_embedding) + (1 - query(bert_weight))* closeness("field", image_sift_encoding)
+  }
+}
+</pre>
+<p>Our search request need to pass both the BERT embedding tensor and the SIFT encoding tensor and we use logical disjunction (OR) to combine the
+two nearest neighbor search operators, our search request becomes: </p>
```

Comment:
Consider splitting sentence in two: "...operators, our..." -> "operators. The search..."